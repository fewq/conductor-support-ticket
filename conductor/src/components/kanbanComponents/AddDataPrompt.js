// import React from "react";
//
// export default ({message}) =>
//  <div className="add-data-prompt action">{`${message}. Type then hit Enter`}</div>;
