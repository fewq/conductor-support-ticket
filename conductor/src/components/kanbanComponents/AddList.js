import React, {PropTypes} from "react";
import Editor from "./Editor";

const AddListProptypes = {
  onClickAddAList : PropTypes.func.isRequired,
  onClickSaveList : PropTypes.func.isRequired,
  itemToEdit : PropTypes.oneOfType([PropTypes.string, PropTypes.number ]).isRequired,
	attributeToEdit  : PropTypes.string.isRequired
};
const AddList = (props) => {
  const showAddListEditor = props.itemToEdit === "board" && props.attributeToEdit === "new list";
  return(
    <div className="list-container">
      <div className="list">
        <div className= { showAddListEditor? "add-a-list-btn-container show-addlist-editor": "add-a-list-btn-container"}>
          {
            showAddListEditor ?
            <Editor
              textareaClass={"add-list"}
              placeholder="Add a list..."
              onClickSave={props.onClickSaveList} /> :
            <div onClick={props.onClickAddAList} className="add-a-list-btn">
              Add a list...
            </div>
          }
        </div>
      </div>
    </div>
  );
};

AddList.propTypes = AddListProptypes;

export default AddList;
