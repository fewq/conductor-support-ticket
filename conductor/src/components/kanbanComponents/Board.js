import React,{PropTypes, Component} from "react";
import HTML5Backend from "react-dnd-html5-backend";
import {DragDropContext} from "react-dnd";

const BoardPropTypes = {
	children : PropTypes.node.isRequired,
};

class Board extends Component {
	render () {
		return(
			<div className="app-board-container">
				<div className="app-board">
					{this.props.children }
				</div>
			</div>
		);
	}
}

Board.propTypes = BoardPropTypes;

const dragDropEnabledBoard = DragDropContext(HTML5Backend)(Board);

export default dragDropEnabledBoard;
