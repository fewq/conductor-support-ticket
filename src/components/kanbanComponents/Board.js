import React, { Component } from "react";
import PropTypes from "prop-types";
import HTML5Backend from "react-dnd-html5-backend";
import { DragDropContext } from "react-dnd";

const BoardPropTypes = {
  children: PropTypes.node.isRequired
};

export class Board extends Component {
  render() {
    return (
      <div className="kanban-board-container">
        <div className="kanban-board">{this.props.children}</div>
      </div>
    );
  }
}

Board.propTypes = BoardPropTypes;

const dragDropEnabledBoard = DragDropContext(HTML5Backend)(Board);

export default dragDropEnabledBoard;
