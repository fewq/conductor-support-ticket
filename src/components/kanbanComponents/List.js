import React, { Component } from "react";
import PropTypes from "prop-types";
import CardContainer from "../../containers/CardContainer";
import Editor from "./Editor";
import { DropTarget } from "react-dnd";
import Types from "./staticTypes";
import axios from "axios";

const DropTargetSpec = {
  drop(props, monitor) {
    const tickets = props.cardsByIds;
    const listId = monitor.getItem().parentListId;
    const ticketIds = props.listCards[listId];
    // update priorities of current list
    for (let i = 0; i < ticketIds.length; i++) {
      const ticket = tickets[ticketIds[i]];
      axios
        .patch("http://localhost:4000/ticket/update/" + ticket.ID, {
          priority: i
        })
        .catch(res => console.log(res));
    }
    console.log("Priorities updated");
  },

  hover(props, monitor, component) {
    const item = monitor.getItem(),
      mouseYPosition = monitor.getClientOffset().y;

    if (item.parentListId === props.id) {
      return;
    }

    const index =
      mouseYPosition < component.rectPosition.top + 32 ? 0 : props.numOfCards; // 32 is the minium height of list-title-div
    props.handleOnMoveCard(item.parentListId, item.index, index);

    monitor.getItem().index = index;
    monitor.getItem().parentListId = props.id;
  }
};

const DropTargetCollect = connect => ({
  connectDropTarget: connect.dropTarget()
});

const ListPropTypes = {
  list: PropTypes.shape({
    id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
    name: PropTypes.string.isRequired
  }).isRequired,
  cards: PropTypes.array.isRequired,
  index: PropTypes.number.isRequired,
  itemToEdit: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
    .isRequired,
  onClickAddACard: PropTypes.func.isRequired,
  onClickSave: PropTypes.func.isRequired,
  attributeToEdit: PropTypes.string.isRequired,
  onClickDeleteList: PropTypes.func.isRequired,
  connectDropTarget: PropTypes.func.isRequired,
  numOfCards: PropTypes.number.isRequired,
  handleOnMoveCard: PropTypes.func.isRequired
};

class List extends Component {
  render() {
    const {
        list,
        connectDropTarget,
        onClickSave,
        onClickAddACard,
        attributeToEdit,
        itemToEdit
      } = this.props,
      showEditor = itemToEdit === list.id && attributeToEdit === "new card";
    return connectDropTarget(
      <div
        className="list-container"
        ref={list =>
          (this.rectPosition = list ? list.getBoundingClientRect() : null)
        }
      >
        <div className="list">
          <div className="action delete-list-icon">
            <i
              className="fa fa-trash"
              aria-hidden="true"
              onClick={this.props.onClickDeleteList}
            />
          </div>
          <h2 className="list-title">{list.name}</h2>
          <div className="cards-container">
            {this.props.cards.map((id, i) => (
              <CardContainer
                key={id}
                id={id}
                index={i}
                parentListId={list.id}
              />
            ))}
            {showEditor && (
              <Editor
                textareaClass={"add-card"}
                placeholder="Add a card..."
                onClickSave={onClickSave}
              />
            )}
          </div>
          {/*!showEditor && (
            <a
              href="#"
              onClick={e => {
                e.stopPropagation();
                e.preventDefault(); // prevent action/render from scrolling to the top of the page everytime
                onClickAddACard();
              }}
              className="add-card-btn"
            >
              Add a card...
            </a>
            )*/}
        </div>
      </div>
    );
  }
}

List.propTypes = ListPropTypes;

const dropEnabledList = DropTarget(
  Types.CARD,
  DropTargetSpec,
  DropTargetCollect
)(List);

export default dropEnabledList;
