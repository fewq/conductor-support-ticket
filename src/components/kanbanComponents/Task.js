import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import Editor from "./Editor";
import {
  showEditor,
  closeCardMenu,
  updateTask,
  toggleTaskDone,
  deleteTask
} from "../../actions/kanban";
import axios from "axios";

const mapStateToProps = ({ domainData, kanbanState }, { id }) => ({
  task: domainData.tasks.byId[id],
  itemToEdit: kanbanState.itemToEdit,
  attributeToEdit: kanbanState.attributeToEdit
});

const mapDispatchToProps = (dispatch, { id }) => ({
  onClickTask: e => {
    e.stopPropagation();
    dispatch(showEditor(id, "task"));
    dispatch(closeCardMenu());
  },
  onClickSave: newVal => {
    dispatch(updateTask(id, newVal));
  },
  onToggleTaskDone: taskId => dispatch(toggleTaskDone(taskId)),
  onClickDeleteTask: (taskId, index, cardId) =>
    dispatch(deleteTask(taskId, index, cardId))
});

const TaskProptypes = {
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  task: PropTypes.shape({
    id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
    name: PropTypes.string.isRequired,
    done: PropTypes.bool.isRequired
  }).isRequired,
  index: PropTypes.number.isRequired,
  cardId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  onClickSave: PropTypes.func.isRequired,
  onClickTask: PropTypes.func.isRequired,
  onClickDeleteTask: PropTypes.func.isRequired,
  onToggleTaskDone: PropTypes.func.isRequired,
  itemToEdit: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
    .isRequired,
  attributeToEdit: PropTypes.string.isRequired
};

const Task = props => {
  const { task, itemToEdit, attributeToEdit } = props,
    onClickSa = newVal => {
      const ID = task.ticketID;
      axios
        .get("http://localhost:4000/ticket/view/" + ID)
        .then(response => {
          const ticket = response.data;
          const newList = ticket.tasks;
          newList[task.index].name = newVal;

          axios
            .patch("http://localhost:4000/ticket/update/" + ID, {
              tasks: newList
            })
            .catch(res => console.log(res));
        })
        .catch(function(error) {
          console.log(error);
        });

      props.onClickSave(newVal);
    },
    onClickTo = () => {
      props.onToggleTaskDone(task.id);
      const ID = task.ticketID;
      axios
        .get("http://localhost:4000/ticket/view/" + ID)
        .then(response => {
          const ticket = response.data;
          const newList = ticket.tasks;
          const cur = task.done;
          newList[task.index].done = !cur;

          axios
            .patch("http://localhost:4000/ticket/update/" + ID, {
              tasks: newList
            })
            .catch(res => console.log(res));
        })
        .catch(function(error) {
          console.log(error);
        });
    },
    onClickDe = () => {
      props.onClickDeleteTask(task.id, props.index, props.cardId);
      const ID = task.ticketID;
      axios
        .get("http://localhost:4000/ticket/view/" + ID)
        .then(response => {
          const ticket = response.data;
          const newList = ticket.tasks;
          const ind = task.index;
          newList.splice(ind);

          axios
            .patch("http://localhost:4000/ticket/update/" + ID, {
              tasks: newList
            })
            .catch(res => console.log(res));
        })
        .catch(function(error) {
          console.log(error);
        });
    };
  const shouldShowEditor = itemToEdit === task.id && attributeToEdit === "task";

  return (
    <li className={`task detail ${shouldShowEditor ? "" : "action"}`}>
      <i
        className={`fa  fa-${
          task.done ? "check-" : ""
        }square-o aria-hidden="true"`}
        onClick={onClickTo}
      />
      {shouldShowEditor ? (
        <div className="editor-wrapper">
          <Editor
            textareaClass={"edit-checklist"}
            shouldShowDelete={true}
            initialValue={task.name}
            onClickSave={onClickSa}
            onClickDelete={onClickDe}
          />
        </div>
      ) : (
        <span
          onClick={props.onClickTask}
          className={`${task.done && "done"} task-name`}
        >
          {task.name}
        </span>
      )}
    </li>
  );
};

Task.propTypes = TaskProptypes;

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Task);
