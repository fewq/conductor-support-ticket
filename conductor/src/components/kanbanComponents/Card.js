// ALLOW findDOMNode because of DND
/* eslint "react/no-find-dom-node": 0 */

import React, {PropTypes, Component} from "react";
import {findDOMNode} from "react-dom";
import TaskList from "./TaskList";
import CardMenu from "./CardMenu";
import Editor from "./Editor";
import Task from "./Task";
import {DragSource, DropTarget} from "react-dnd";
import Types from "../staticTypes";

let initialState; // initialState before beginDrag DnD

const dragSourceSpec = {
	beginDrag : ({id, index, parentListId, curState}) => ({
		id,
		index,
		parentListId,
		curState
	}),

	isDragging : (props, monitor) =>
		monitor.getItem().id === props.id
 ,

	// endDrag (props) {
	// 	// if (!monitor.didDrop()) {props.replacePlaceholderWithCurDraggingCard(props.curState); return;}
	// 	// props.replacePlaceholderWithCurDraggingCard(props.curState);
	// }
},

dragSourceCollect = function dragSourceCollect (connect, monitor) {
	return {
		isDragging : monitor.isDragging(),
		connectDragSource : connect.dragSource(),
		connectDragPreview : connect.dragPreview(),
		dragItem : monitor.getItem()
	};
};

const cardDropTargetSpec = {
  hover(props, monitor, component) {
    const item   = monitor.getItem(),
					hoverId = props.id,
					dragId = item.id,
					dragIndex = item.index,
					hoverIndex = props.index;

    // Don't replace items with themselves
    if (dragId === hoverId) {
      return;
    }

    // Determine rectangle on screen
    const hoverBoundingRect = findDOMNode(component).getBoundingClientRect();

    // Get vertical middle
    const hoverMiddleY = (hoverBoundingRect.bottom - hoverBoundingRect.top) / 2;

    // Determine mouse position
    const clientOffset = monitor.getClientOffset();

    // Get pixels to the top
    const hoverClientY = clientOffset.y - hoverBoundingRect.top;

    // Only perform the move when the mouse has crossed half of the items height
    // When dragging downwards, only move when the cursor is below 50%
    // When dragging upwards, only move when the cursor is above 50%

    // Dragging downwards
    if (dragIndex < hoverIndex && hoverClientY < hoverMiddleY) {
      return;
    }

    // Dragging upwards
    if (dragIndex > hoverIndex && hoverClientY > hoverMiddleY) {
      return;
    }

    // Time to actually perform the action
    props.handleOnSortCard(hoverId, hoverIndex, dragId, dragIndex);

    // Note: we're mutating the monitor item here!
    // Generally it's better to avoid mutations,
    // but it's good here for the sake of performance
    // to avoid expensive index searches.
    monitor.getItem().index = hoverIndex;
  }
};

const cardDropTargetCollect = (connect, monitor) => ({
	isOver : monitor.isOver(),
	connectDropTarget : connect.dropTarget()
});


class Card extends Component{
	componentDidMount() {
		if (initialState !== undefined) {
			this.props.onSetState(initialState);
			initialState = undefined;
		}
	}

	render () {
		const {
			card, onClickDeleteCard, onShowDetails,connectDragSource,
			connectDropTarget, showEditor, isSelectedCard,
			attributeToEdit, updateCard, itemToEdit,isDragging,dragItem
		} = this.props,
		isItemToEdit = itemToEdit === card.id,
		onClickSave = (value)	=>{updateCard(attributeToEdit, value);};

		// const { dragItem} = this.props;
		//
		if (isDragging) {
			initialState = dragItem.curState;
		}
		return connectDragSource(
			connectDropTarget(
				<div className="card ">
					<div className="menu-icon action" >
						<i className="fa fa-ellipsis-v"
								aria-hidden="true"
								onClick={(e) => {
										this.props.onClickMenu(e.target);
										e.stopPropagation();
									}
								}>
						</i>
					</div>
					{
						this.props.shouldShowCardMenu &&
						<CardMenu
							hasDescription    = {!!card.description.length}
							onClickAddATask   = {(e)=>{onShowDetails(); showEditor("new task"); e.stopPropagation();}}
							onClickEditTitle  = {(e)=>{onShowDetails(); showEditor("title"); e.stopPropagation(); }}
							menuPosition    = {this.props.menuPosition}
							onClickDeleteCard  = {onClickDeleteCard}
							onClickEditDescription = {(e)=>{onShowDetails(); showEditor( "description"); e.stopPropagation(); }}
						/>
					}
					<div className="card-title-container">
						{

							isItemToEdit &&  attributeToEdit === "title" ?
							<Editor
								textareaClass = {"edit-checklist"}
								initialValue   = {card.title}
								onClickSave   = {onClickSave}
							/> :

							<div className="card-title action" onClick={this.props.onToggleShowDetails}>
								<i className={this.props.shouldShowDetails?"fa fa-caret-down":"fa fa-caret-right"} aria-hidden="true"></i>
								<span className="title">{card.title}</span>
							</div>
						}
					</div>
					{
						this.props.shouldShowDetails === true ?
						<div className="card-details">
							<div className="description">
								{
									isItemToEdit && attributeToEdit === "description" ?
									<Editor
										textareaClass = {"edit-checklist"}
										placeholder   = "Add a description..."
										initialValue   = {card.description}
										onClickSave   = {onClickSave}
									/> :

								<div onClick={(e)=>{showEditor("description"); e.stopPropagation();}}
										className={`action  ${card.description ? "card-description" : "add-description-prompt"}`}>
										{card.description ? card.description : "Add a description..."}
									</div>
								}
							</div>
							<TaskList>
								{
									card.tasks.map((taskId, i) =>
										<Task
											id = {taskId}
											key = {taskId}
											index = {i}
											cardId = {card.id}
										/>)
								}
							</TaskList>

						<div className={`add-task-wrapper ${isSelectedCard &&  attributeToEdit === "addTask" ? "active" : ""}`}>
								{
									isItemToEdit && attributeToEdit === "new task" ?
									<Editor
										textareaClass = {"edit-checklist"}
										placeholder   = "Add a task..."
										onClickSave   = {(taskName)=>{this.props.onAddATask(card.id, taskName);}}
									/> :
									<div onClick={(e)=>{showEditor( "new task"); e.stopPropagation();}}>Add a task...</div>
								}
							</div>

						</div> : undefined
					}
					<div className={isDragging ? "placeholder" : ""}></div>
				</div>
			)
		);
	}
}

Card.propTypes = {
	card  : PropTypes.shape({
		id    : PropTypes.oneOfType([PropTypes.string, PropTypes.number ]).isRequired,
		title : PropTypes.string.isRequired,
		tasks : PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.string, PropTypes.number ])).isRequired,
		description : PropTypes.string.isRequired
	}).isRequired,
	index : PropTypes.number.isRequired,
	parentListId : PropTypes.oneOfType([PropTypes.string, PropTypes.number ]).isRequired,
	shouldShowDetails : PropTypes.bool.isRequired,
	shouldShowCardMenu : PropTypes.bool.isRequired,
	currentEditorValue : PropTypes.string,
	itemToEdit : PropTypes.oneOfType([PropTypes.string, PropTypes.number ]).isRequired,
	attributeToEdit : PropTypes.string.isRequired,
	isSelectedCard  : PropTypes.bool.isRequired,
	onClickMenu 		: PropTypes.func.isRequired,
	curTaskEditing  : PropTypes.oneOfType([PropTypes.string, PropTypes.number ])/*.isRequired*/,
	showEditor 	: PropTypes.func.isRequired,
	onShowDetails : PropTypes.func.isRequired,
	updateCard 	: PropTypes.func.isRequired,
	menuPosition      : PropTypes.object.isRequired,
	onAddATask 				: PropTypes.func.isRequired,
	onToggleShowDetails : PropTypes.func.isRequired,
	onClickDeleteCard      : PropTypes.func.isRequired,
	curState : PropTypes.object.isRequired,// we want to preserve card state while/after Dn*/,
	isDragging : PropTypes.bool,
	connectDragSource : PropTypes.func.isRequired,
	onSetState : PropTypes.func.isRequired,
	connectDropTarget : PropTypes.func.isRequired,
	dragItem : PropTypes.object,
	isOver :  PropTypes.bool.isRequired,
	handleOnSortCard : PropTypes.func.isRequired
};

const dropableCard = DropTarget(Types.CARD, cardDropTargetSpec, cardDropTargetCollect)(Card);
const draggableCard = DragSource(Types.CARD, dragSourceSpec, dragSourceCollect)(dropableCard);

export default draggableCard;
