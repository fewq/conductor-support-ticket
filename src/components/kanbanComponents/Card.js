// ALLOW findDOMNode because of DND
/* eslint "react/no-find-dom-node": 0 */

import React, { Component } from "react";
import { Link } from "react-router-dom";
import { findDOMNode } from "react-dom";
import PropTypes from "prop-types";
import TaskList from "./TaskList";
import CardMenu from "./CardMenu";
import Editor from "./Editor";
import Task from "./Task";
import { DragSource, DropTarget } from "react-dnd";
import Types from "./staticTypes";
import axios from "axios";
import jwtDecode from "jwt-decode";

let initialState; // initialState before begin DnD

const dragSourceSpec = {
    beginDrag: ({ id, index, parentListId, curState }) => ({
      id,
      index,
      parentListId,
      curState
    }),

    isDragging: (props, monitor) => monitor.getItem().id === props.id,

    endDrag(props, monitor) {
      let newStatusToAdmin;
      const newParent = monitor.getItem().parentListId;
      switch (newParent) {
        case "0":
          newStatusToAdmin = "Pending Admin";
          break;
        case "1":
          newStatusToAdmin = "Pending BA";
          break;
        case "2":
          newStatusToAdmin = "Pending Developers";
          break;
        case "3":
          newStatusToAdmin = "Pending Client";
          break;
        case "4":
          newStatusToAdmin = "Resolved";
          break;
        case "5":
          newStatusToAdmin = "Deleted";
          break;
        default:
          break;
      }

      const ticketID = props.card.ID;

      // update status if status changed
      if (props.card.statusToAdmin != newStatusToAdmin) {
        // local change
        props.updateCard("statusToAdmin", newStatusToAdmin);
        props.updateCard("notified", false);

        // backend change
        axios
          .patch("http://localhost:4000/ticket/update/" + ticketID, {
            statusToAdmin: newStatusToAdmin,
            notified: false
          })
          .then(res => {
            console.log(
              "Changed status of ticket to admin, and notified status"
            );
            console.log(res.data.statusToAdmin);
            console.log("Notified: " + res.data.notified);
          })
          .catch(res => console.log(res));
      }
    }
  },
  dragSourceCollect = function dragSourceCollect(connect, monitor) {
    return {
      isDragging: monitor.isDragging(),
      connectDragSource: connect.dragSource(),
      connectDragPreview: connect.dragPreview(),
      dragItem: monitor.getItem()
    };
  };

const cardDropTargetSpec = {
  hover(props, monitor, component) {
    const item = monitor.getItem(),
      hoverId = props.id,
      dragId = item.id,
      dragIndex = item.index,
      hoverIndex = props.index;

    // Don't replace items with themselves
    if (dragId === hoverId) {
      return;
    }

    // Determine rectangle on screen
    const hoverBoundingRect = findDOMNode(component).getBoundingClientRect();

    // Get vertical middle
    const hoverMiddleY = (hoverBoundingRect.bottom - hoverBoundingRect.top) / 2;

    // Determine mouse position
    const clientOffset = monitor.getClientOffset();

    // Get pixels to the top
    const hoverClientY = clientOffset.y - hoverBoundingRect.top;

    // Only perform the move when the mouse has crossed half of the items height
    // When dragging downwards, only move when the cursor is below 50%
    // When dragging upwards, only move when the cursor is above 50%

    // Dragging downwards
    if (dragIndex < hoverIndex && hoverClientY < hoverMiddleY) {
      return;
    }

    // Dragging upwards
    if (dragIndex > hoverIndex && hoverClientY > hoverMiddleY) {
      return;
    }

    // Time to actually perform the action
    props.handleOnSortCard(hoverId, hoverIndex, dragId, dragIndex);

    // Note: we're mutating the monitor item here!
    // Generally it's better to avoid mutations,
    // but it's good here for the sake of performance
    // to avoid expensive index searches.
    monitor.getItem().index = hoverIndex;
  }
};

const cardDropTargetCollect = (connect, monitor) => ({
  isOver: monitor.isOver(),
  connectDropTarget: connect.dropTarget()
});

class Card extends Component {
  componentDidMount() {
    if (initialState !== undefined) {
      this.props.onSetState(initialState);
      initialState = undefined;
    }
  }

  render() {
    const {
        card,
        onClickDeleteCard,
        onShowDetails,
        connectDragSource,
        connectDropTarget,
        showEditor,
        isSelectedCard,
        attributeToEdit,
        updateCard,
        itemToEdit,
        isDragging,
        dragItem
      } = this.props,
      isItemToEdit = itemToEdit === card.id,
      onClickSave = value => {
        updateCard(attributeToEdit, value);
        let newValues;
        switch (attributeToEdit) {
          case "title":
            newValues = {
              title: value
            };
            break;
          case "description":
            newValues = {
              description: value
            };
            break;
        }

        axios
          .patch("http://localhost:4000/ticket/update/" + card.ID, newValues)
          .then(res => {
            console.log("Changed status of ticket to client");
            console.log(res.data);
          })
          .catch(res => console.log(res));
      },
      onClickSend = value => {
        const title = card.title;
        const status = card.status;
        const message = value;
        const link = card.ID;
        const email = card.email;
        const target = "client";
        if (status != "Pending Admin") {
          axios.post("/api/notify", {
            email,
            title,
            status,
            message,
            link,
            target
          });
        } else {
          console.log("Email not sent");
        }
        this.props.updateCard("notified", true);

        let payload = {
          attendedBy: jwtDecode(localStorage.getItem("id_token")).email,
          ticketId: link,
          prevStatusToClient: status, // CHANGE
          statusToClient: card.statusToAdmin,
          statusToAdmin: card.statusToAdmin,
          dateOfUpdate: new Date(),
          comments: message
        };

        axios
          .post("http://localhost:4000/status/add", payload)
          .then(res => {
            console.log(
              "Adding new ticket status update with the following info:"
            );
            console.log(res.data);
          })
          .catch(res => console.log(res));

        /*
        const updatedDescription =
          card.description +
          "\n\nAdmin's message: (" +
          new Date().toString() +
          ")\n" +
          message;
        console.log(updatedDescription);
        this.props.updateCard("description", updatedDescription);
*/
        axios
          .patch("http://localhost:4000/ticket/update/" + link, {
            notified: true
          })
          .catch(res => console.log(res));
      };

    // const { dragItem} = this.props;
    //
    if (isDragging) {
      initialState = dragItem.curState;
    }
    return connectDragSource(
      connectDropTarget(
        <div className="card ">
          <div className="menu-icon action">
            <i
              className="fa fa-ellipsis-v"
              aria-hidden="true"
              onClick={e => {
                this.props.onClickMenu(e.target);
                console.log(e.target);
                e.stopPropagation();
              }}
            />
          </div>
          {this.props.shouldShowCardMenu && (
            <CardMenu
              hasDescription={!!card.description.length}
              onClickAddATask={e => {
                onShowDetails();
                showEditor("new task");
                e.stopPropagation();
              }}
              onClickEditTitle={e => {
                onShowDetails();
                showEditor("title");
                e.stopPropagation();
              }}
              menuPosition={this.props.menuPosition}
              onClickDeleteCard={e => {
                e.stopPropagation();
                onClickDeleteCard(card.ID);
              }}
              onClickEditDescription={e => {
                onShowDetails();
                showEditor("description");
                e.stopPropagation();
              }}
              onClickNotify={() => {
                this.props.onClickNotify();
              }}
              onClickToggleNotify={() => {
                this.props.onClickToggleNotify(card.notified, card.ID);
              }}
              status={card.statusToAdmin}
            />
          )}
          <div className="card-title-container">
            {isItemToEdit && attributeToEdit === "title" ? (
              <Editor
                textareaClass={"edit-checklist"}
                initialValue={card.title}
                onClickSave={onClickSave}
              />
            ) : (
              <div
                className="card-title action"
                onClick={this.props.onToggleShowDetails}
              >
                <i
                  className={
                    this.props.shouldShowDetails
                      ? "fa fa-caret-down"
                      : "fa fa-caret-right"
                  }
                  aria-hidden="true"
                />
                <span className="title">{card.title}</span>
              </div>
            )}
          </div>
          {isItemToEdit && attributeToEdit === "message" && (
            <Editor
              textareaClass={"edit-checklist"}
              placeholder="Add a message..."
              initialValue={"Default message"}
              buttonText={"msg"}
              onClickSave={onClickSend}
            />
          )}

          {this.props.shouldShowDetails === true ? (
            <div className="card-details">
              <div className="description">
                {isItemToEdit && attributeToEdit === "description" ? (
                  <Editor
                    textareaClass={"edit-checklist"}
                    placeholder="Add a description..."
                    initialValue={card.description}
                    onClickSave={onClickSave}
                  />
                ) : (
                  <div
                    onClick={e => {
                      showEditor("description");
                      e.stopPropagation();
                    }}
                    className={`action  ${
                      card.description
                        ? "card-description"
                        : "add-description-prompt"
                    }`}
                  >
                    {card.description
                      ? card.description
                      : "Add a description..."}
                  </div>
                )}
              </div>

              <TaskList>
                {card.tasks.map((taskId, i) => (
                  <Task id={taskId} key={taskId} index={i} cardId={card.id} />
                ))}
              </TaskList>

              <div
                className={`add-task-wrapper ${
                  isSelectedCard && attributeToEdit === "addTask"
                    ? "active"
                    : ""
                }`}
              >
                {isItemToEdit && attributeToEdit === "new task" ? (
                  <Editor
                    textareaClass={"edit-checklist"}
                    placeholder="Add a task..."
                    onClickSave={taskName => {
                      this.props.onAddATask(card.id, taskName);
                    }}
                  />
                ) : (
                  <div
                    onClick={e => {
                      showEditor("new task");
                      e.stopPropagation();
                    }}
                  >
                    Add a task...
                  </div>
                )}
              </div>
              <Link
                to={{
                  pathname: `/view/${card.ID}`,
                  state: {
                    ticket: card.ticket
                  }
                }}
                className="card-view"
                id="card-view-details"
              >
                View details
              </Link>
            </div>
          ) : (
            undefined
          )}
          <div className={isDragging ? "placeholder" : ""} />
          {card.statusToAdmin != "Pending Admin" &&
            (card.notified ? (
              <div className="card-notified"> Client Notified </div>
            ) : (
              <div className="card-not-notified"> Client Not Notified </div>
            ))}
        </div>
      )
    );
  }
}

Card.propTypes = {
  card: PropTypes.shape({
    id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
    ID: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
    title: PropTypes.string.isRequired,
    tasks: PropTypes.arrayOf(
      PropTypes.oneOfType([PropTypes.string, PropTypes.number])
    ).isRequired,
    description: PropTypes.string.isRequired,
    status: PropTypes.string.isRequired
  }).isRequired,
  index: PropTypes.number.isRequired,
  parentListId: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
    .isRequired,
  shouldShowDetails: PropTypes.bool.isRequired,
  shouldShowCardMenu: PropTypes.bool.isRequired,
  currentEditorValue: PropTypes.string,
  itemToEdit: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
    .isRequired,
  attributeToEdit: PropTypes.string.isRequired,
  isSelectedCard: PropTypes.bool.isRequired,
  onClickMenu: PropTypes.func.isRequired,
  curTaskEditing: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]) /*.isRequired*/,
  showEditor: PropTypes.func.isRequired,
  onShowDetails: PropTypes.func.isRequired,
  updateCard: PropTypes.func.isRequired,
  menuPosition: PropTypes.object.isRequired,
  onAddATask: PropTypes.func.isRequired,
  onToggleShowDetails: PropTypes.func.isRequired,
  onClickNotify: PropTypes.func.isRequired,
  onClickToggleNotify: PropTypes.func.isRequired,
  onClickDeleteCard: PropTypes.func.isRequired,
  curState: PropTypes.object.isRequired, // we want to preserve card state while/after Dn*/,
  isDragging: PropTypes.bool,
  connectDragSource: PropTypes.func.isRequired,
  onSetState: PropTypes.func.isRequired,
  connectDropTarget: PropTypes.func.isRequired,
  dragItem: PropTypes.object,
  isOver: PropTypes.bool.isRequired,
  handleOnSortCard: PropTypes.func.isRequired
};

const dropableCard = DropTarget(
  Types.CARD,
  cardDropTargetSpec,
  cardDropTargetCollect
)(Card);
const draggableCard = DragSource(Types.CARD, dragSourceSpec, dragSourceCollect)(
  dropableCard
);

export default draggableCard;
