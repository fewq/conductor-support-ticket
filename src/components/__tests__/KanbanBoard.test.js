import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import { Board } from "../kanbanComponents/Board";

test("Renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<Board />, div);
  ReactDOM.unmountComponentAtNode(div);
});

test("snapshot testing", () => {
  const tree = renderer.create(<Board />).toJSON();
  expect(tree).toMatchSnapshot();
});
