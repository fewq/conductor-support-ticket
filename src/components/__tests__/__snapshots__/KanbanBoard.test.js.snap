// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot testing 1`] = `
<div
  className="kanban-board-container"
>
  <div
    className="kanban-board"
  />
</div>
`;
